platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/inputs/undeclared/MethodReferenceInstrumentationIntegrationTest.groovy

```diff
@@ -94,6 +95,181 @@ class MethodReferenceInstrumentationIntegrationTest extends AbstractConfiguratio
+                        return String.valueOf(checkExists(new MyFile(input)));
+                    }
+
+                    static boolean checkExists(MyFile myFile) {
+                        Supplier<Boolean> supplier = myFile::isFile;
+                        return supplier.get();
+                    }
+                }
+            """
+        }
+
+        testDirectory.create {
+            createDir("buildSrc") {
+                file("src/main/java/MethodRefInputsCls.java") << classTemplate("class", "MethodRefInputsCls")
+                file("src/main/java/MethodRefInputsInterface.java") << classTemplate("interface", "MethodRefInputsInterface")
+            }
+        }
+
+        buildFile """
+            tasks.register("echo") {
+                def value1 = MethodRefInputsCls.readInputWithReference(${input.expr})
+                def value2 = MethodRefInputsInterface.readInputWithReference(${input.expr})
+                doLast {
+                    println("value1 = \$value1")
+                    println("value2 = \$value2")
+                }
+            }
+        """
+
+        when:
+        configurationCacheRun("echo")
+
+        then:
+        outputContains("value1 = ${input.expectedValue}")
+        outputContains("value2 = ${input.expectedValue}")
+
+        problems.assertResultHasProblems(result) {
+            withInput("Build file 'build.gradle': ${input.expectedInput}")
+        }
+    }
+
+    def "does not instrument bound method reference when receiver is a subtype that overrides the method"() {
+        given:
+        def input = fileEntry()
+        def classTemplate = { ownerKind, ownerName -> """
+                import java.io.*;
+                import java.nio.file.*;
+                import java.util.*;
+                import java.util.function.*;
+
+                public $ownerKind $ownerName {
+                    public static class MyFile extends File {
+                        public MyFile(String path) { super(path); }
+
+                        @Override public boolean isFile() { return false; }  // does not call super intentionally.
+                    }
+
+                    public static String readInputWithReference(String input) throws Exception {
+                        return String.valueOf(checkExists(new MyFile(input)));
+                    }
+
+                    static boolean checkExists(MyFile myFile) {
+                        Supplier<Boolean> supplier = myFile::isFile;
+                        return supplier.get();
+                    }
+                }
+            """
+        }
+
+        testDirectory.create {
+            createDir("buildSrc") {
+                file("src/main/java/MethodRefInputsCls.java") << classTemplate("class", "MethodRefInputsCls")
+                file("src/main/java/MethodRefInputsInterface.java") << classTemplate("interface", "MethodRefInputsInterface")
```

Comment:
❔ What is difference when testing interface vs class here?

Also will some test fail due to static method in interface for Java8 tests?

Reply:
> ❔ What is difference when testing interface vs class here?

This isn't a focus of this particular test case, but as it is an end-to-end test, I decided to test both. There is a subtle difference in the method reference of the generated bridge method.

> ~Also will some test fail due to static method in interface for Java8 tests?~ I guess no

No, we have tests doing that already. Java 8 doesn't support `private` methods in interfaces, but just statics are fine.